File: src/lib/engine/direction.ts
import type { Analysis, CastEntry, Emotion, Segment, SynthesisStyle } from "@/lib/types";
import { NARRATOR_ID } from "@/lib/types";
import { clamp } from "@/lib/utils";
import { languageName } from "./lang";

const EMOTION_PROSODY: Record<Emotion, { pitch: number; rate: number; volume: number }> = {
  neutral: { pitch: 0, rate: 0, volume: 0 },
  happy: { pitch: 5, rate: 6, volume: 0 },
  sad: { pitch: -5, rate: -10, volume: -8 },
  angry: { pitch: 3, rate: 6, volume: 10 },
  fearful: { pitch: 6, rate: 8, volume: -4 },
  surprised: { pitch: 8, rate: 5, volume: 4 },
  excited: { pitch: 7, rate: 10, volume: 5 },
  whisper: { pitch: -3, rate: -8, volume: -30 },
  calm: { pitch: -2, rate: -5, volume: -2 },
  serious: { pitch: -3, rate: -3, volume: 0 },
  sarcastic: { pitch: 2, rate: -2, volume: 0 },
  tender: { pitch: -1, rate: -7, volume: -8 },
  shouting: { pitch: 8, rate: 8, volume: 20 },
};

const EMOTION_DIRECTION: Record<Emotion, string> = {
  neutral: "naturally, in character",
  happy: "with warmth and a smile in the voice",
  sad: "with sadness, softly and a little slower",
  angry: "with barely controlled anger, sharp and forceful",
  fearful: "with fear, breathless and trembling",
  surprised: "with genuine surprise",
  excited: "with bright excitement and energy",
  whisper: "in a hushed whisper",
  calm: "calmly and evenly",
  serious: "gravely and seriously",
  sarcastic: "with dry sarcasm",
  tender: "tenderly and gently",
  shouting: "shouting loudly",
};

export function styleFor(segment: Segment, analysis: Analysis, entry: CastEntry): SynthesisStyle {
  const isNarrator = segment.speaker === NARRATOR_ID;
  const damp = isNarrator ? 0.4 : 1;
  const mod = EMOTION_PROSODY[segment.emotion] ?? EMOTION_PROSODY.neutral;
  const lang = languageName(segment.lang ?? analysis.language);
  let instructions: string;
  if (isNarrator) {
    const tone = analysis.narrator.tone || "warm, measured storyteller";
    instructions = `You are the narrator of an audiobook: ${tone}. Speak ${lang}. Read with an engaging storytelling cadence${
      segment.emotion !== "neutral" ? `, subtly ${EMOTION_DIRECTION[segment.emotion]}` : ""
    }.`;
  } else {
    const c = analysis.characters.find((x) => x.id === segment.speaker);
    const who = c
      ? `${c.name}, ${c.age === "adult" ? "an adult" : `a ${c.age}`} ${c.gender === "neutral" ? "person" : c.gender}${
          c.personality.length ? ` who is ${c.personality.join(", ")}` : ""
        }`
      : "a character in the story";
    const manner = c?.speakingStyle ? ` Speaking style: ${c.speakingStyle}.` : "";
    const timbre = c?.voiceDescription ? ` Voice: ${c.voiceDescription}.` : "";
    instructions = `Perform as ${who}.${timbre}${manner} Speak ${lang}. Deliver this line ${EMOTION_DIRECTION[segment.emotion]}.`;
  }
  return {
    pitch: clamp(entry.pitch + mod.pitch * damp, -40, 40),
    rate: clamp(entry.rate + mod.rate * damp, -45, 50),
    volume: clamp(mod.volume * damp, -40, 30),
    emotion: segment.emotion,
    instructions,
  };
}

File: android/app/src/main/java/com/nigixmosha/app/engine/Direction.java
package com.nigixmosha.app.engine;

import com.nigixmosha.app.engine.model.Types;

import java.util.HashMap;
import java.util.Map;

public final class Direction {
    private static final Map<String, double[]> PROSODY = new HashMap<>();
    private static final Map<String, String> MANNER = new HashMap<>();

    static {
        PROSODY.put("neutral", new double[]{0, 0, 0});
        PROSODY.put("happy", new double[]{5, 6, 0});
        PROSODY.put("sad", new double[]{-5, -10, -8});
        PROSODY.put("angry", new double[]{3, 6, 10});
        PROSODY.put("fearful", new double[]{6, 8, -4});
        PROSODY.put("surprised", new double[]{8, 5, 4});
        PROSODY.put("excited", new double[]{7, 10, 5});
        PROSODY.put("whisper", new double[]{-3, -8, -30});
        PROSODY.put("calm", new double[]{-2, -5, -2});
        PROSODY.put("serious", new double[]{-3, -3, 0});
        PROSODY.put("sarcastic", new double[]{2, -2, 0});
        PROSODY.put("tender", new double[]{-1, -7, -8});
        PROSODY.put("shouting", new double[]{8, 8, 20});

        MANNER.put("neutral", "naturally, in character");
        MANNER.put("happy", "with warmth and a smile in the voice");
        MANNER.put("sad", "with sadness, softly and a little slower");
        MANNER.put("angry", "with barely controlled anger, sharp and forceful");
        MANNER.put("fearful", "with fear, breathless and trembling");
        MANNER.put("surprised", "with genuine surprise");
        MANNER.put("excited", "with bright excitement and energy");
        MANNER.put("whisper", "in a hushed whisper");
        MANNER.put("calm", "calmly and evenly");
        MANNER.put("serious", "gravely and seriously");
        MANNER.put("sarcastic", "with dry sarcasm");
        MANNER.put("tender", "tenderly and gently");
        MANNER.put("shouting", "shouting loudly");
    }

    private Direction() {
    }

    public static Types.SynthesisStyle styleFor(Types.Segment segment, Types.Analysis analysis, Types.CastEntry entry) {
        boolean isNarrator = Types.NARRATOR_ID.equals(segment.speaker);
        double damp = isNarrator ? 0.4 : 1;
        double[] mod = PROSODY.containsKey(segment.emotion) ? PROSODY.get(segment.emotion) : PROSODY.get("neutral");
        String manner = MANNER.containsKey(segment.emotion) ? MANNER.get(segment.emotion) : MANNER.get("neutral");
        String lang = Lang.languageName(segment.lang != null ? segment.lang : analysis.language);
        String instructions;
        if (isNarrator) {
            String tone = analysis.narrator.tone == null || analysis.narrator.tone.isEmpty() ? "warm, measured storyteller" : analysis.narrator.tone;
            instructions = "You are the narrator of an audiobook: " + tone + ". Speak " + lang
                    + ". Read with an engaging storytelling cadence"
                    + (!"neutral".equals(segment.emotion) ? ", subtly " + manner : "") + ".";
        } else {
            Types.CastCharacter c = analysis.character(segment.speaker);
            String who;
            if (c != null) {
                String agePart = "adult".equals(c.age) ? "an adult" : "a " + c.age;
                String genderPart = "neutral".equals(c.gender) ? "person" : c.gender;
                String traits = c.personality.isEmpty() ? "" : " who is " + String.join(", ", c.personality);
                who = c.name + ", " + agePart + " " + genderPart + traits;
            } else {
                who = "a character in the story";
            }
            String styleNote = c != null && c.speakingStyle != null && !c.speakingStyle.isEmpty() ? " Speaking style: " + c.speakingStyle + "." : "";
            String timbre = c != null && c.voiceDescription != null && !c.voiceDescription.isEmpty() ? " Voice: " + c.voiceDescription + "." : "";
            instructions = "Perform as " + who + "." + timbre + styleNote + " Speak " + lang + ". Deliver this line " + manner + ".";
        }
        Types.SynthesisStyle style = new Types.SynthesisStyle();
        style.pitch = Util.clamp(entry.pitch + mod[0] * damp, -40, 40);
        style.rate = Util.clamp(entry.rate + mod[1] * damp, -45, 50);
        style.volume = Util.clamp(mod[2] * damp, -40, 30);
        style.emotion = segment.emotion;
        style.instructions = instructions;
        return style;
    }
}
